pkg/yurthub/proxy/multiplexer/multiplexerlist.go

```diff
@@ -0,0 +1,159 @@
+	scope, err := sp.getReqScope(gvr)
+	if err != nil {
+		util.Err(errors.Wrapf(err, "failed to get request scope"), w, r)
+		return
+	}
+
+	listOpts, err := sp.decodeListOptions(r, scope)
+	if err != nil {
+		util.Err(errors.Wrapf(err, "failed to decode list options, url: %v", r.URL), w, r)
+		return
+	}
+
+	storageOpts, err := sp.storageOpts(listOpts, gvr)
+	if err != nil {
+		util.Err(err, w, r)
+		return
+	}
+
+	obj, err := sp.listObject(r, gvr, storageOpts)
+	if err != nil {
+		util.Err(err, w, r)
+		return
+	}
+
+	util.WriteObject(http.StatusOK, obj, w, r)
+}
+
+func (sp *multiplexerProxy) listObject(r *http.Request, gvr *schema.GroupVersionResource, storageOpts *kstorage.ListOptions) (runtime.Object, error) {
+	rc, _, err := sp.cacheManager.ResourceCache(gvr)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get resource cache")
+	}
+
+	obj, err := sp.newListObject(gvr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to new list object")
+	}
+
+	key, err := sp.getCacheKey(r, storageOpts)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get cache key")
+	}
+
+	if err := rc.GetList(r.Context(), key, *storageOpts, obj); err != nil {
+		return nil, errors.Wrapf(err, "failed to get list from cache")
+	}
+
+	if obj, err = sp.filterListObject(obj, sp.filterMgr.FindObjectFilters(r)); err != nil {
+		return nil, errors.Wrapf(err, "failed to filter list object")
+	}
+
+	return obj, nil
+}
+
+func (sp *multiplexerProxy) newListObject(gvr *schema.GroupVersionResource) (runtime.Object, error) {
+	rcc, err := sp.cacheManager.ResourceCacheConfig(gvr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get resource cache config")
+	}
+
+	return rcc.NewListFunc(), nil
+}
+
+func (sp *multiplexerProxy) getCacheKey(r *http.Request, storageOpts *kstorage.ListOptions) (string, error) {
+	if ns := sp.getNamespace(r); len(ns) > 0 {
+		return sp.getNamespaceScopedCacheKey(r, storageOpts)
+	}
+
+	return sp.getClusterScopedCacheKey(r, storageOpts)
+}
+
+func (sp *multiplexerProxy) getNamespaceScopedCacheKey(r *http.Request, storageOpts *kstorage.ListOptions) (string, error) {
+	ctx := request.WithNamespace(r.Context(), sp.getNamespace(r))
+
+	if name, ok := storageOpts.Predicate.MatchesSingle(); ok {
+		return registry.NamespaceKeyFunc(ctx, "", name)
+	}
+
+	return registry.NamespaceKeyRootFunc(ctx, ""), nil
+}
+
+func (sp *multiplexerProxy) getNamespace(r *http.Request) string {
+	requestInfo, ok := request.RequestInfoFrom(r.Context())
+	if !ok {
+		return ""
+	}
+	return requestInfo.Namespace
+}
+
+func (sp *multiplexerProxy) getClusterScopedCacheKey(r *http.Request, storageOpts *kstorage.ListOptions) (string, error) {
+	if name, ok := storageOpts.Predicate.MatchesSingle(); ok {
+		return registry.NoNamespaceKeyFunc(r.Context(), "", name)
+	}
+
+	return "", nil
+}
+
+func (sp *multiplexerProxy) filterListObject(obj runtime.Object, filter filter.ObjectFilter) (runtime.Object, error) {
+	if filter == nil {
```

Comment:
filter may not be nil because of it is a filter.ObjectFilter. so please use `yurtutil.IsNil()` func to verify.

Reply:
done